### PR TITLE
XWIKI-19750: Allow macros to have several categories

### PR DESCRIPTION
**Pull request by manuelleduc**
Depends on: https://github.com/xwiki/xwiki-rendering/pull/216
Jira: https://jira.xwiki.org/browse/XWIKI-19750

* Update the wiki macro xar files to use the new `defaultCategories` property
* Update the java macro to call `setDefaultCategories` (allowing `setDefaultCategory` to be moved to legacy on `xwiki-rendering`).
* Introduction of `xwiki-platform-legacy-rendering-wikimacro` and `xwiki-platform-legacy-rendering-wikimacro-store`
* Add an around aspect on `DefaultWikiMacroFactory.getDefaultCategories` to fallback to `defaultCategory` if no categories are found by the default implementation
* Move `WikiMacroConstants.MACRO_DEFAULT_CATEGORY_PROPERTY` to legacy
* Update `WikiMacroClassDocumentInitializer` to initialize `defaultCategories` instead of `defaultCategory`
* Add `xwiki-platform-legacy-rendering-wikimacro-store` to `xwiki-platform-distribution-war-legacydependencies`
* Move `org.xwiki.rendering.macro.wikibridge.WikiMacroDescriptor.Builder#defaultCategory` to legacy
* Move `WikiMacroDescriptor.getDefaultCategory` to legacy

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-legacy/xwiki-platform-legacy-rendering/xwiki-platform-legacy-rendering-wikimacro/xwiki-platform-legacy-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/LegacyWikiMacroConstants.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * @since 14.6RC1
+ */
+public interface LegacyWikiMacroConstants
+{
+    /**
+     * Constant for representing default macro category property.
+     *
+     * @deprecated since 14.6RC1, replaced by {@link WikiMacroConstants#MACRO_DEFAULT_CATEGORIES_PROPERTY}
+     */
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-legacy/xwiki-platform-legacy-rendering/xwiki-platform-legacy-rendering-wikimacro/xwiki-platform-legacy-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/LegacyWikiMacroConstants.java:35]: This one is a copy paste but indeed, does not hurt to add it, fixing.

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java [hunk 2]
```diff
@@ -154,9 +154,10 @@ private WikiMacro buildMacro(XWikiDocument doc)
             this.logger.debug("Incomplete macro definition in [{}], macro description is empty", documentReference);
         }
 
+        List<String> macroDefaultCategories = getDefaultCategories(macroDefinition);
         // Verify default macro category.
-        if (StringUtils.isEmpty(macroDefaultCategory)) {
-            macroDefaultCategory = null;
+        if (macroDefaultCategories == null || macroDefaultCategories.isEmpty()) {
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java:159]: `CollectionUtils.inEmpty` ?

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java [hunk 3]
```diff
@@ -365,4 +372,9 @@ public boolean isAllowed(DocumentReference documentReference, WikiMacroVisibilit
 
         return isAllowed;
     }
+
+    private List<String> getDefaultCategories(BaseObject macroDefinition)
+    {
+        return (List<String>) macroDefinition.getListValue(MACRO_DEFAULT_CATEGORIES_PROPERTY);
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java:378]: Would using `#getSetValue` make more sense here ?

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java [hunk 4]
```diff
@@ -84,12 +85,13 @@ public Builder description(String description)
...
 
         /**
-         * @param defaultCategory the macro default category
-         * @return this builder
+         * @param defaultCategories the macro default categories
+         * @return this build
+         * @since 14.6RC1
          */
-        public Builder defaultCategory(String defaultCategory)
+        public Builder defaultCategories(Set<String> defaultCategories)
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java:92]: Note that you return whatever Set was provided in input, and it might be a modifiable Set (for example it's the case with your implementation of wiki macros builder) which is not great for #getDefaultCategories.

File: xwiki-platform-distribution/xwiki-platform-distribution-war-legacydependencies/pom.xml [hunk 5]
```diff
@@ -228,6 +232,37 @@
          </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-legacy-rendering-wikimacro-api</artifactId>
+      <version>${project.version}</version>
+      <exclusions>
+        <exclusion>
+          <artifactId>xwiki-platform-rendering-wikimacro-api</artifactId>
```
Comment (tmortagne) [inline xwiki-platform-distribution/xwiki-platform-distribution-war-legacydependencies/pom.xml:241]: This one is not a transitive dependency so no need to exclude it.

File: xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml [hunk 6]
```diff
@@ -233,18 +338,30 @@
...
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.StringClass</classType>
+      </type>
     </class>
     <property>
       <defaultValue/>
     </property>
     <property>
-      <description>The reference to the file in an attachment archive (e.g. "attach:space.page@attachment/my/file.txt")</description>
+      <description>The reference to the file in an attachment archive (e.g. "[[attachment/my/file.txt")&gt;&gt;attach:space.page@attachment/my/file.txt")]]</description>
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml:355]: Was that change on purpose ?

File: xwiki-platform-core/xwiki-platform-legacy/xwiki-platform-legacy-rendering/xwiki-platform-legacy-rendering-wikimacro/xwiki-platform-legacy-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/LegacyWikiMacroConstants.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.xwiki.rendering.wikimacro.internal;
+
+/**
+ * Class containing fields moved to legacy from {@link WikiMacroConstants}.
+ *
+ * @version $Id$
+ * @since 14.6RC1
+ */
+public interface LegacyWikiMacroConstants
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-legacy/xwiki-platform-legacy-rendering/xwiki-platform-legacy-rendering-wikimacro/xwiki-platform-legacy-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/LegacyWikiMacroConstants.java:28]: Shouldn't this whole interface we annotated as deprecated ?

File: xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml (hunk 6 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml:355]: For each xar file change, I edited a running wiki to copy the value of `defaultCategory` to `defaultCategories` and remove the deprecated field and did a xar export. So the change is the result of this process. But indeed it looks strange, I'll check what happened!

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java (hunk 4 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java:92]: Do you suggest to wrap the Set passed in parameter into an immutable Set internally?

Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java:92]: Ideally you should copy the input in a new immutable Set to be sure that there can not be any leak.

Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java:92]: Depends of what we use, say we choose to use `java.util.Collections#unmodifiableSet`, a Set used in parameter so we can't take any collection.

File: xwiki-platform-distribution/xwiki-platform-distribution-war-legacydependencies/pom.xml (hunk 5 above)
Comment (manuelleduc) [reply xwiki-platform-distribution/xwiki-platform-distribution-war-legacydependencies/pom.xml:241]: After verification this exclusion has no effect and as the dependency is not drawn from `xwiki-platform-legacy-rendering-wikimacro-api`, removing.

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java (hunk 3 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-store/src/main/java/org/xwiki/rendering/wikimacro/internal/DefaultWikiMacroFactory.java:378]: Good catch thanks, I'll change that.

File: xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java (hunk 4 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-wikimacro/xwiki-platform-rendering-wikimacro-api/src/main/java/org/xwiki/rendering/macro/wikibridge/WikiMacroDescriptor.java:92]: java.util.Collections#unmodifiableSet does not created a new Set, it just wraps it.

File: xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml (hunk 6 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-ui/src/main/resources/Macros/VFSTreeMacro.xml]: Indeed I did another export and the description stays the same, reverting this change.
